### PR TITLE
support native staking v2 76

### PR DESCRIPTION
**Pull request by lzxm160**
work on #76 

need to add tests when testnet deployed with latest iotex-core

### CODE REVIEW

File: iotex/interfaces.go [hunk 1]
```diff
@@ -67,6 +67,8 @@ type AuthedClient interface {
 	ClaimReward(value *big.Int) ClaimRewardCaller
 	DeployContract(data []byte) DeployContractCaller
 	Account() account.Account
+	// staking related
+	StakingCaller(interface{}) StakingCaller
```
Comment (dustinxie) [inline iotex/interfaces.go:71]: move to before Account()
and better name it Staking() StakingCaller for consistency

File: iotex/client.go [hunk 2]
```diff
@@ -62,6 +62,15 @@ func (c *authedClient) DeployContract(data []byte) DeployContractCaller {
 
 func (c *authedClient) Account() account.Account { return c.account }
 
+// StakingCaller
```
Comment (dustinxie) [inline iotex/client.go:65]: moved to before Account

File: iotex/callers.go [hunk 3]
```diff
@@ -34,6 +34,12 @@ type sendActionCaller struct {
 	nonce    *uint64
 }
 
+// reclaim type to differentiate unstake and withdraw
+type reclaim struct {
+	action     interface{}
+	isWithdraw bool
+}
+
```
Comment (dustinxie) [inline iotex/callers.go:42]: this can go to caller_staking.go?

File: iotex/interfaces.go (hunk 1 above)
Comment (lzxm160) [reply iotex/interfaces.go:71]: sure

File: iotex/callers.go (hunk 3 above)
Comment (lzxm160) [reply iotex/callers.go:42]: use this to differentiate unstake and withdraw,it's in the same pakage with caller_staking.go

File: iotex/interfaces.go [hunk 4]
```diff
@@ -66,6 +66,8 @@ type AuthedClient interface {
 	Transfer(to address.Address, value *big.Int) TransferCaller
 	ClaimReward(value *big.Int) ClaimRewardCaller
 	DeployContract(data []byte) DeployContractCaller
+	// staking related
+	Staking(interface{}) StakingCaller
```
Comment (yutongp) [inline iotex/interfaces.go:70]: I  would say this is not a clear api. it is not clear what `interface` should be put as input here. as you can see, all other apis have clean input.

I would suggest api as following:
```
....
Staking() StakingCaller
Candidate() CandidateCaller
```
```
type StakingCaller interface {
  Create(candidateName string, amount *big.Int, duration uint32, autoStake bool) StakeCreateCaller
  Unstake(bucketIndex uint64) StakeUnstakeCaller
  Withdraw(bucketIndex uint64) StakeWithdrawCaller
  AddDeposit(index uint64, amount *big.Int) StakeAddDepositCaller
  ChangeCandidate(candName string, bucketIndex uint64) StakeChangeCandidateCaller
  Transfer(voterAddress address.Address, bucketIndex uint64) StakeTransferCaller
  Restake(index uint64, duration uint32, autoStake bool) RestakeCaller
}

type CandidateCaller interface {
  Register(
	  name, operatorAddr, rewardAddr  address.Address,
	  duration uint32,
	  autoStake bool) CandidateRegisterCaller // owner address is optional by default is sender's address
 Update(name string, operatorAddr, rewardAddr address.Address) CandidateUpdateCaller
}
```
  
}
```

Comment (lzxm160) [reply iotex/interfaces.go]: done with this comment,pls check

File: iotex/interfaces.go [hunk 5]
```diff
@@ -103,3 +106,74 @@ type Contract interface {
 type ReadOnlyContract interface {
 	Read(method string, args ...interface{}) ReadContractCaller
 }
+
+// StakingCaller is used to perform a staking call.
+type StakingCaller interface {
+	Create(candidateName string, amount *big.Int, duration uint32, autoStake bool) StakeCreateCaller
```
Comment (dustinxie) [inline iotex/interfaces.go:112]: i think for simplicity, can use Create() StakingCaller, no need to define another StakeCreateCaller
same for others, no need to define separate StakeUnstakeCaller etc.

to do create stake: c.Staking().Create(, , ,).Call()
to do unstake: c.Staking().Unstake(, , ,).Call()
there is no confusion

File: iotex/interfaces.go [hunk 6]
```diff
@@ -103,3 +106,74 @@ type Contract interface {
...
+// StakingBase is for staking
+type StakingBase interface {
+	SendActionCaller
+	SetGasPrice(*big.Int) StakingBase
+	SetGasLimit(uint64) StakingBase
+	SetNonce(uint64) StakingBase
+	SetPayload([]byte) StakingBase
+}
+
+// StakeCreateCaller is for stake create call.
```
Comment (dustinxie) [inline iotex/interfaces.go:136]: we don't need these 8 separate definition

File: iotex/caller_staking.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+type (
+	stakingCaller struct {
+		account account.Account
+		api     iotexapi.APIServiceClient
+	}
+	candidateCaller struct {
+		account account.Account
+		api     iotexapi.APIServiceClient
+	}
+	stakingBase struct {
```
Comment (dustinxie) [inline iotex/caller_staking.go:32]: can we combine this with stakingCaller, see comments below

File: iotex/interfaces.go [hunk 8]
```diff
@@ -103,3 +106,30 @@ type Contract interface {
...
+	SetGasLimit(uint64) StakingCaller
+	SetNonce(uint64) StakingCaller
+	SetPayload([]byte) StakingCaller
+	Create(candidateName string, amount *big.Int, duration uint32, autoStake bool) StakingCaller
+	Unstake(bucketIndex uint64) StakingCaller
+	Withdraw(bucketIndex uint64) StakingCaller
+	AddDeposit(index uint64, amount *big.Int) StakingCaller
+	ChangeCandidate(candName string, bucketIndex uint64) StakingCaller
+	StakingTransfer(voterAddress address.Address, bucketIndex uint64) StakingCaller
+	Restake(index uint64, duration uint32, autoStake bool) StakingCaller
```
Comment (yutongp) [inline iotex/interfaces.go:123]: it is confusing to have these functions at same level.
what does it mean if I do
`StakingCaller.Unstake(...).Withdraw(...).AddDeposit(...).Call`?

Comment (yutongp) [reply iotex/interfaces.go:123]:
```
type StakingCaller interface {
  Create(candidateName string, amount *big.Int, duration uint32, autoStake bool) StakingAPICaller
  Unstake(bucketIndex uint64) StakingAPICaller
  Withdraw(bucketIndex uint64) StakingAPICaller
  AddDeposit(index uint64, amount *big.Int) StakingAPICaller
  ChangeCandidate(candName string, bucketIndex uint64) StakingAPICaller
  Transfer(voterAddress address.Address, bucketIndex uint64) StakingAPICaller
  Restake(index uint64, duration uint32, autoStake bool) StakingAPICaller
}
```
you can do this to reduce dup code if you want

Comment (lzxm160) [reply iotex/interfaces.go:123]: good point，I‘ll change it’

**Pull request merged**